openedx/openedx-platform
# feat: adds Content Tagging

## Description

Adds the models, APIs, and permissions/rules for `openedx.features.content_tagging`, which builds on the `oel_tagging` app changes made by https://github.com/openedx/openedx-learning/pull/62

These changes are the first step towards allow Course Authors to tag content objects (courses, units, libraries) in edx-platform, but none of these changes are user-facing yet.

## Supporting information

Part of https://github.com/openedx/modular-learning/issues/63

## Testing instructions

This change should be 100% covered by unit tests.

But since there's no public-facing APIs yet, if you want to do any further verification, you can test the functionality from the devstack and shell. 

1. Get the `master` [openedx/devstack](https://github.com/openedx/devstack) up and running.
2. Pull this branch into edx-platform
3. Install dependencies and update migrations by running `paver install_prereqs` in the LMS shell (from `devstack` dir, run `make lms-shell`) and CMS shell (`make studio-shell`).
4. Run the migrations added here by running `./manage.py lms migrate` in the LMS shell.
5. Create file `cms/envs/private.py` which contains the following. You may need to restart Studio for these changes to take effect: `make dev.restart-devserver.studio`
   ```
   from .devstack import FEATURES
   # Org-based course authoring restrictions
   FEATURES['ENABLE_CREATOR_GROUP'] = True
   FEATURES['STUDIO_REQUEST_EMAIL'] = 'your@email.com'
   ```
6. Go to the Studio home page and sign up a new user: http://localhost:18010/
7. Expand the "Becoming a Course Creator in Studio" box to request course creation access (this is optional -- an "unrequested" entry is created even if you don't request one, and you can approve that below.)
8. Login to Studio Django Admin in another window as a superuser (http://localhost:18010/admin, `edx@example.com` / `edx`)
9. Create a new Organization, e.g. with Name: OpenCraft, Short Name: OC
10. Navigate to [Course creators|http://localhost:18010/admin/course_creators/coursecreator/] and mark the request `granted`, and select the OpenCraft organization as the only org you can create courses for.

From here, you'll have to play around in the shell, because we only have Django Admin capabilities for ContentTaxonomies right now, and you have to be global staff to access Django Admin, which doesn't let us test out our org restrictions.

For example:

```
# make studio-shell
root@lms:/edx/app/edxapp/edx-platform# ./manage.py cms shell

from django.contrib.auth import get_user_model
# user you created
org_user = get_user_model().objects.get(username='openedx')
# demo users automatically created
learner = get_user_model().objects.get(username='honor')
staff_user = get_user_model().objects.get(username='staff')

from openedx.features.content_tagging import models, rules, api

from organizations.models import Organization
oc = Organization.objects.get(short_name="OC")
org_taxonomy = api.create_taxonomy(name="OpenCraft taxonomy", org_owners=[oc])
org_taxonomy
# <Taxonomy object (2)>
assert org_user.has_perm('oel_tagging.add_taxonomy', org_taxonomy)
assert org_user.has_perm('oel_tagging.change_taxonomy', org_taxonomy)
assert not learner.has_perm('oel_tagging.add_taxonomy')
assert not learner.has_perm('oel_tagging.change_taxonomy', org_taxonomy)
assert not learner.has_perm('oel_tagging.add_taxonomy', org_taxonomy)
```

## Deadline

None

## Other information

Depends on https://github.com/openedx/openedx-learning/pull/62 and https://github.com/openedx/openedx-learning/pull/65

Data migrations added here can be easily rolled back.

Course creator groups added by https://github.com/openedx/edx-platform/pull/26616.

### Author to do before merge:

- [ ] Update dependency: https://github.com/openedx/edx-platform/pull/32661/files#r1253990343
